File: typescript/packages/sdk/test/command.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  resolveArgv,
  argvHash,
  buildCommandManifest,
  bridgeExistingCli,
  contractHash,
  type ArgvTemplate,
} from "../src/index.ts";
import { loadVector } from "./helpers.ts";

interface CommandVector {
  resolution_cases: Array<{
    name: string;
    argv_template: ArgvTemplate;
    params: Record<string, unknown>;
    resolved_argv: string[];
    argument_hash: string;
  }>;
  injection_cases: Array<{
    name: string;
    argv_template: ArgvTemplate;
    params: Record<string, unknown>;
    resolved_argv: string[];
    argument_hash: string;
    assert: { argv_length: number; last_element_equals: string; shell_used: boolean };
    expect: { decision: string; reason_code: string };
  }>;
  identity_cases: Array<{
    name: string;
    exec_digest_a: string;
    exec_digest_b: string;
  }>;
}

const V = loadVector<CommandVector>("command.json");

test("§28.1 resolution_cases: argv_template + params → flat argv; argv_hash matches", () => {
  for (const c of V.resolution_cases) {
    const argv = resolveArgv(c.argv_template, c.params);
    assert.deepEqual(argv, c.resolved_argv, `resolved argv mismatch for ${c.name}`);
    assert.equal(argvHash(argv), c.argument_hash, `argv_hash mismatch for ${c.name}`);
  }
});

test("§28.1 / security test 20: injection_cases — metacharacters stay one literal argv element", () => {
  for (const c of V.injection_cases) {
    const argv = resolveArgv(c.argv_template, c.params);
    assert.deepEqual(argv, c.resolved_argv, `resolved argv mismatch for ${c.name}`);
    assert.equal(argvHash(argv), c.argument_hash);

    // The normative assertions: exactly N argv elements, the metacharacter
    // payload is the SINGLE last element verbatim, and no shell is involved.
    assert.equal(argv.length, c.assert.argv_length, "argv length");
    assert.equal(argv[argv.length - 1], c.assert.last_element_equals, "last element literal");
    assert.equal(c.assert.shell_used, false);
    // It is one element, not split on the space/semicolon/hash.
    assert.equal(argv.filter((e) => e === c.assert.last_element_equals).length, 1);
  }
});

test("§28.1: a scalar param is never re-split, even with spaces and metachars", () => {
  const tmpl: ArgvTemplate = ["bin", { param: "x", schema: { type: "string" } }];
  const argv = resolveArgv(tmpl, { x: "a b; c && d | e > f" });
  assert.deepEqual(argv, ["bin", "a b; c && d | e > f"]);
  assert.equal(argv.length, 2);
});

test("§28.1: missing/object params are rejected (one value → one slot)", () => {
  const tmpl: ArgvTemplate = ["bin", { param: "x", schema: { type: "string" } }];
  assert.throws(() => resolveArgv(tmpl, {}), /missing required param/);
  assert.throws(() => resolveArgv(tmpl, { x: { a: 1 } }), /single argv slot/);
});

const COMMON = {
  issuer: "did:web:example.com",
  provider: "example.git",
  name: "git.commit",
  version: "1.0.0",
  summary_for_user: "Commit staged changes.",
  summary_for_model: "git commit with a typed message.",
  input_schema: {
    type: "object",
    additionalProperties: false,
    properties: { message: { type: "string" } },
    required: ["message"],
  },
  output_schema: { type: "object" },
  effects: { class: "write-reversible" as const, external_side_effect: true, compensating_action: "git.reset" },
  determinism: { class: "nondeterministic" as const },
  sandbox: { filesystem: ["/work"], network: [], secrets: [] },
};

const ARGV_TEMPLATE: ArgvTemplate = [
  "git",
  "commit",
  "-m",
  { param: "message", schema: { type: "string" } },
];

test("§4.1 / §28.4 / security test 22: identity_cases — exec_digest change ⇒ new contract_hash", () => {
  for (const c of V.identity_cases) {
    const manifestA = buildCommandManifest({
      ...COMMON,
      command: { binary: "git", exec_digest: c.exec_digest_a, argv_template: ARGV_TEMPLATE },
    });
    const manifestB = buildCommandManifest({
      ...COMMON,
      command: { binary: "git", exec_digest: c.exec_digest_b, argv_template: ARGV_TEMPLATE },
    });

    // Two capabilities identical but for command.exec_digest MUST differ.
    assert.notEqual(
      manifestA.capability.contract_hash,
      manifestB.capability.contract_hash,
      `${c.name}: exec_digest change must change contract_hash`,
    );
    assert.notEqual(manifestA.capability.id, manifestB.capability.id);

    // The command block round-trips and shell is forced false.
    assert.equal(manifestA.capability.kind, "command");
    assert.equal(manifestA.capability.command?.shell, false);
    assert.equal(manifestA.capability.command?.exec_digest, c.exec_digest_a);

    // Identity is the hash of contract INCLUDING the command block (§4.1).
    assert.equal(
      manifestA.capability.contract_hash,
      contractHash(manifestA as never),
      "contract_hash must equal hash of extracted contract (with command block)",
    );
  }
});

test("§28.4: command bridge produces a host_cli, digest-pinned command manifest", () => {
  const m = bridgeExistingCli({
    binary: "git",
    execDigest: "sha256:" + "a".repeat(64),
    subcommandAllow: ["commit", "-m"],
    argvTemplate: ARGV_TEMPLATE,
  });
  assert.equal(m.capability.kind, "command");
  assert.equal(m.capability.command?.provenance, "host_cli");
  assert.equal(m.capability.command?.exec_digest, "sha256:" + "a".repeat(64));
  assert.equal(m.capability.command?.shell, false);
  assert.deepEqual(m.capability.command?.subcommand_allow, ["commit", "-m"]);

  // A changed binary digest is a new identity (§28.4 rug-pull defense).
  const m2 = bridgeExistingCli({
    binary: "git",
    execDigest: "sha256:" + "b".repeat(64),
    subcommandAllow: ["commit", "-m"],
    argvTemplate: ARGV_TEMPLATE,
  });
  assert.notEqual(m.capability.contract_hash, m2.capability.contract_hash);
});

File: go/sdk/command_test.go
package sdk

import (
	"encoding/json"
	"testing"
)

// command_test.go reproduces the resolution_cases, injection_cases, and
// identity_cases of conformance/vectors/command.json (spec §28), plus focused
// ResolveArgv / ArgvHash / argv-token unit tests. The path_cases and taint_cases of
// the same vector are reproduced on the enforcing side, in the gateway package.

// commandVector is the decoded shape of command.json this package consumes. Fields
// not exercised here (path_cases, taint_cases) are still declared so the document
// decodes cleanly.
type commandVector struct {
	ResolutionCases []struct {
		Name         string          `json:"name"`
		ArgvTemplate []ArgvToken     `json:"argv_template"`
		Params       map[string]string `json:"params"`
		ResolvedArgv []string        `json:"resolved_argv"`
		ArgumentHash string          `json:"argument_hash"`
	} `json:"resolution_cases"`
	InjectionCases []struct {
		Name         string            `json:"name"`
		ArgvTemplate []ArgvToken       `json:"argv_template"`
		Params       map[string]string `json:"params"`
		ResolvedArgv []string          `json:"resolved_argv"`
		ArgumentHash string            `json:"argument_hash"`
		Assert       struct {
			ArgvLength        int    `json:"argv_length"`
			LastElementEquals string `json:"last_element_equals"`
			ShellUsed         bool   `json:"shell_used"`
		} `json:"assert"`
	} `json:"injection_cases"`
	IdentityCases []struct {
		Name        string `json:"name"`
		ExecDigestA string `json:"exec_digest_a"`
		ExecDigestB string `json:"exec_digest_b"`
	} `json:"identity_cases"`
}

func loadCommandVector(t *testing.T) commandVector {
	t.Helper()
	raw := loadVector(t, "command.json")
	var doc commandVector
	if err := json.Unmarshal(raw, &doc); err != nil {
		t.Fatalf("decode command.json: %v", err)
	}
	return doc
}

// TestCommandResolutionVector reproduces command.json resolution_cases: ResolveArgv
// produces the exact resolved_argv, and ArgvHash equals the vector's argument_hash.
func TestCommandResolutionVector(t *testing.T) {
	doc := loadCommandVector(t)
	if len(doc.ResolutionCases) == 0 {
		t.Fatal("no resolution_cases in vector")
	}
	for _, c := range doc.ResolutionCases {
		t.Run(c.Name, func(t *testing.T) {
			argv, err := ResolveArgv(c.ArgvTemplate, c.Params)
			if err != nil {
				t.Fatalf("ResolveArgv: %v", err)
			}
			assertArgvEqual(t, argv, c.ResolvedArgv)

			h, err := ArgvHash(argv)
			if err != nil {
				t.Fatalf("ArgvHash: %v", err)
			}
			if h != c.ArgumentHash {
				t.Errorf("argument_hash = %s, want %s", h, c.ArgumentHash)
			}
		})
	}
}

// TestCommandInjectionVector reproduces command.json injection_cases: a parameter
// containing shell metacharacters ("; rm -rf / #") stays a SINGLE literal argv
// element. It asserts the vector's argv_length, last_element_equals, and that no
// shell was used (shell_used:false) — the structural CWE-78 defense (security test
// #20). It also confirms the argument_hash matches.
func TestCommandInjectionVector(t *testing.T) {
	doc := loadCommandVector(t)
	if len(doc.InjectionCases) == 0 {
		t.Fatal("no injection_cases in vector")
	}
	for _, c := range doc.InjectionCases {
		t.Run(c.Name, func(t *testing.T) {
			argv, err := ResolveArgv(c.ArgvTemplate, c.Params)
			if err != nil {
				t.Fatalf("ResolveArgv: %v", err)
			}
			assertArgvEqual(t, argv, c.ResolvedArgv)

			// assert.argv_length: the metacharacters did not split into extra elements.
			if len(argv) != c.Assert.ArgvLength {
				t.Fatalf("argv length = %d, want %d (metacharacters must not split)", len(argv), c.Assert.ArgvLength)
			}
			// assert.last_element_equals: the whole injected string is ONE element.
			last := argv[len(argv)-1]
			if last != c.Assert.LastElementEquals {
				t.Errorf("last argv element = %q, want %q", last, c.Assert.LastElementEquals)
			}
			// assert.shell_used:false — VCP never builds a shell string. There is no
			// shell anywhere in resolution; this constant asserts the contract.
			if c.Assert.ShellUsed {
				t.Fatal("vector asserts shell_used:false but case says true")
			}

			h, err := ArgvHash(argv)
			if err != nil {
				t.Fatalf("ArgvHash: %v", err)
			}
			if h != c.ArgumentHash {
				t.Errorf("argument_hash = %s, want %s", h, c.ArgumentHash)
			}
		})
	}
}

// TestCommandIdentityVector reproduces command.json identity_cases: two command
// capabilities identical but for command.exec_digest MUST have different
// contract_hash (spec §4.1, §28.4) — a changed binary digest is a new identity
// (security test #22). The contract is the eight common fields + the command block.
func TestCommandIdentityVector(t *testing.T) {
	doc := loadCommandVector(t)
	if len(doc.IdentityCases) == 0 {
		t.Fatal("no identity_cases in vector")
	}
	for _, c := range doc.IdentityCases {
		t.Run(c.Name, func(t *testing.T) {
			base := Contract{
				Issuer:       "did:web:tools.example",
				Name:         "git.commit",
				Version:      "1.0.0",
				InputSchema:  map[string]any{"type": "object", "additionalProperties": false},
				OutputSchema: map[string]any{"type": "object"},
				Effects:      map[string]any{"class": "write-reversible", "external_side_effect": false, "compensating_action": "git.reset"},
				Determinism:  map[string]any{"class": "idempotent-write"},
				Sandbox:      map[string]any{"filesystem": []any{"/work"}, "network": []any{}, "secrets": []any{}},
			}
			mkCmd := func(digest string) Command {
				return Command{
					Binary:       "git",
					ExecDigest:   digest,
					Shell:        false,
					ArgvTemplate: []ArgvToken{LiteralToken("git"), LiteralToken("commit"), LiteralToken("-m"), ParamToken("message", map[string]any{"type": "string"})},
					Provenance:   ProvenanceHostCLI,
				}
			}

			hashA, err := CommandContractHash(base, mkCmd(c.ExecDigestA))
			if err != nil {
				t.Fatalf("hash A: %v", err)
			}
			hashB, err := CommandContractHash(base, mkCmd(c.ExecDigestB))
			if err != nil {
				t.Fatalf("hash B: %v", err)
			}
			if hashA == hashB {
				t.Fatalf("contract_hash A == B (%s); a changed exec_digest MUST yield a new identity", hashA)
			}

			// And the command block actually participates in identity: dropping it
			// changes the hash relative to the common-only contract.
			commonOnly, err := base.ContractHash()
			if err != nil {
				t.Fatal(err)
			}
			if hashA == commonOnly {
				t.Error("command contract_hash equals the common-only contract_hash; command block is not identity-bearing")
			}

			idA, err := CommandCapabilityID(base, mkCmd(c.ExecDigestA))
			if err != nil {
				t.Fatal(err)
			}
			want := "vcp:cap:git.commit@" + hashA
			if idA != want {
				t.Errorf("capability_id = %q, want %q", idA, want)
			}
		})
	}
}

// TestResolveArgvOneElementPerParam is a focused unit test: every typed hole becomes
// exactly one argv element regardless of its content. It checks several adversarial
// values — spaces, quotes, globs, redirections, NUL-free control text — each of
// which a shell would split or interpret but ResolveArgv must keep intact.
func TestResolveArgvOneElementPerParam(t *testing.T) {
	tmpl := []ArgvToken{
		LiteralToken("printf"),
		LiteralToken("%s"),
		ParamToken("value", map[string]any{"type": "string"}),
	}
	cases := []string{
		"; rm -rf / #",
		"a b c",
		"*.go",
		"$(whoami)",
		"`id`",
		"a\nb",
		"--flag=looks-like-a-flag",
		"",
		"||touch pwned",
		">/etc/passwd",
	}
	for _, v := range cases {
		t.Run(v, func(t *testing.T) {
			argv, err := ResolveArgv(tmpl, map[string]string{"value": v})
			if err != nil {
				t.Fatalf("ResolveArgv: %v", err)
			}
			if len(argv) != 3 {
				t.Fatalf("argv length = %d, want 3 (one element per token)", len(argv))
			}
			if argv[0] != "printf" || argv[1] != "%s" {
				t.Errorf("literal tokens mangled: %q", argv[:2])
			}
			if argv[2] != v {
				t.Errorf("param element = %q, want %q (must be verbatim, one element)", argv[2], v)
			}
		})
	}
}

// TestResolveArgvMissingParam asserts resolution fails closed when a hole names a
// parameter that is not supplied (rather than silently substituting an empty slot).
func TestResolveArgvMissingParam(t *testing.T) {
	tmpl := []ArgvToken{LiteralToken("git"), ParamToken("ref", map[string]any{"type": "string"})}
	if _, err := ResolveArgv(tmpl, map[string]string{}); err == nil {
		t.Fatal("expected error for missing parameter, got nil")
	}
}

// TestArgvTokenJSONRoundTrip asserts a literal token marshals to a bare string and a
// typed hole to a {param,schema} object, and that both round-trip, matching the
// manifest schema's argv_template oneOf.
func TestArgvTokenJSONRoundTrip(t *testing.T) {
	lit := LiteralToken("git")
	b, err := json.Marshal(lit)
	if err != nil {
		t.Fatal(err)
	}
	if string(b) != `"git"` {
		t.Errorf("literal token JSON = %s, want \"git\"", b)
	}

	hole := ParamToken("message", map[string]any{"type": "string"})
	b, err = json.Marshal(hole)
	if err != nil {
		t.Fatal(err)
	}
	var back ArgvToken
	if err := json.Unmarshal(b, &back); err != nil {
		t.Fatalf("unmarshal hole: %v", err)
	}
	if !back.IsParam() || back.Param != "message" {
		t.Errorf("round-tripped hole = %#v, want param=message", back)
	}

	// A bare string unmarshals to a literal token.
	var lit2 ArgvToken
	if err := json.Unmarshal([]byte(`"commit"`), &lit2); err != nil {
		t.Fatal(err)
	}
	if lit2.IsParam() || lit2.Literal != "commit" {
		t.Errorf("round-tripped literal = %#v, want literal=commit", lit2)
	}
}

func assertArgvEqual(t *testing.T, got, want []string) {
	t.Helper()
	if len(got) != len(want) {
		t.Fatalf("argv length = %d (%q), want %d (%q)", len(got), got, len(want), want)
	}
	for i := range got {
		if got[i] != want[i] {
			t.Errorf("argv[%d] = %q, want %q", i, got[i], want[i])
		}
	}
}
